codalab-service.py

```diff
@@ -0,0 +1,387 @@
+        'mysql_port': 'CODALAB_MYSQL_PORT',
+        'ssl_cert_file': 'CODALAB_SSL_CERT_FILE',
+        'ssl_key_file': 'CODALAB_SSL_KEY_FILE',
+    }
+
+    @staticmethod
+    def _get_parser():
+        parser = argparse.ArgumentParser(description="Manages your local CodaLab worksheets back-end deployment using docker-compose")
+        subparsers = parser.add_subparsers(dest='command', description='Command to run')
+
+        # SUBCOMMANDS
+
+        start_cmd = subparsers.add_parser('start', help='Start a CodaLab backend instance')
+        logs_cmd = subparsers.add_parser('logs', help='View logs for existing CodaLab instance')
+        test_cmd = subparsers.add_parser('test', help='Run tests against an existing CodaLab instance')
+        build_cmd = subparsers.add_parser('build', help='Build CodaLab docker images using the local codebase')
+        run_cmd = subparsers.add_parser('run', help='Run a command inside a service container')
+
+        stop_cmd = subparsers.add_parser('stop', help='Stop any existing CodaLab backend instance')
+        down_cmd = subparsers.add_parser('down', help='Bring down any existing CodaLab backend instance')
+        restart_cmd = subparsers.add_parser('restart', help='Restart any existing CodaLab backend instance')
+
+        #  BUILD SETTINGS
+
+        for cmd in [build_cmd, start_cmd]:
+            cmd.add_argument('--version', '-v', type=str, help='CodaLab version to use for building and deployment', default=argparse.SUPPRESS)
+            cmd.add_argument('--dev', action='store_true', help='If specified use dev versions of images', default=argparse.SUPPRESS)
+            cmd.add_argument('--push', action='store_true', help='If specified push the images to Dockerhub', default=argparse.SUPPRESS)
+            cmd.add_argument('--docker-user', type=str, help='DockerHub username to push images from', default=argparse.SUPPRESS)
+            cmd.add_argument('--docker-pwd', type=str, help='DockerHub password to push images from', default=argparse.SUPPRESS)
+
+        build_cmd.add_argument('image', default=argparse.SUPPRESS, help='Images to build', choices=['bundleserver', 'frontend', 'worker', 'default-cpu', 'default-gpu', 'all'], nargs='?')
+
+        #  DEPLOYMENT SETTINGS
+        for cmd in [start_cmd, stop_cmd, restart_cmd, down_cmd, logs_cmd]:
+            cmd.add_argument('--instance-name', type=str, help='Docker compose name to use for instance', default=argparse.SUPPRESS)
+
+        start_cmd.add_argument('--build-locally', '-b', action='store_true', help='If specified build VERSION using local code.', default=argparse.SUPPRESS)
+        start_cmd.add_argument('--test-build', '-t', action='store_true', help='If specified run tests on the build.', default=argparse.SUPPRESS)
+        start_cmd.add_argument('--user-compose-file', type=str, help='If specified path to a user-defined Docker compose file that overwrites the defaults', default=argparse.SUPPRESS)
```

Comment:
why not `--docker-compose-file`?

Reply:
There are already quite a few docker compose files that configure the service, this argument is for situations where you have advanced customizations that go beyond what this script can do. If you specify a user docker compose file, any field in it overrides what's in the service setup.

In that sense it's a docker compose file that allows the user to override the provided docker-compose files. Calling it just `docker-compose-file` didn't seem to convey that but maybe we can change it to `docker-compose-override-file` or something?